resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DbReader.scala

```diff
@@ -64,19 +63,31 @@ object DbReader {
         .stream
         .transact(xa)
 
-    // Return all clusters that has no non "DELETED" | `ERROR` non-default nodepool
-    // TODO: Add grace period
-    override def getK8sClustersToDelete: Stream[F, KubernetesClusterId] =
-      sql"""SELECT kc.id, kc.location, kc.clusterName
-            FROM KUBERNETES_CLUSTER kc 
-            WHERE NOT EXISTS (
-              SELECT *
-                FROM KUBERNETES_CLUSTER kc 
-                LEFT JOIN NODEPOOL as np on np.clusterId=kc.id
-              WHERE np.status != "DELETED" and np.status !="ERROR" and np.isDefault = 0
-            );
+    // Return all non-deleted clusters with non-default nodepools that have apps that were all deleted
+    // or errored outside the grace period (1 hour)
+    //
+    // We are excluding clusters with only default nodepools running on them so we do not remove batch-pre-created clusters.
+    // We are calculating the grace period for cluster deletion assuming that the following are valid proxies for an app's last activity:
+    //    1. destroyedDate for deleted apps
+    //    2. dateAccessed for error'ed apps
+    override def getKubernetesClustersToDelete: Stream[F, KubernetesClusterToRemove] =
+      sql"""SELECT DISTINCT kc.id, kc.googleProject
+            FROM KUBERNETES_CLUSTER AS kc
+            LEFT JOIN NODEPOOL AS np ON kc.id = np.clusterId
```

Comment:
I'm wondering if we could remove these LEFT JOINs and boil the interval checks into the `NOT EXISTS` subquery…

Reply:
```
SELECT kc.id, kc.googleProject
FROM KUBERNETES_CLUSTER kc
WHERE
  kc.status != "DELETED" AND
  NOT EXISTS (
    SELECT *
    FROM NODEPOOL n
    LEFT JOIN APP a ON n.id = a.nodepoolId
    WHERE kc.id = n.clusterId AND 
      ((n.isDefault = 0 AND a.status != "DELETED" AND a.status != "ERROR") OR
        (app.destroyedDate > now() - INTERVAL 1 HOUR) OR
        (app.dateAccessed > now() INTERVAL 1 HOUR)));
```

Reply:
Does that work? This way the subquery checks for apps that are active OR have been recently deleted/accessed

Reply:
Great idea! I think it'll work. I just want to test it a bit more. My dev DB got into a weird state that I need to fix first.

Reply:
Slightly modified version of that query (note the updates within the inner `WHERE` clause below) seems to work.

```sql
            SELECT kc.id, kc.googleProject
            FROM KUBERNETES_CLUSTER kc
            WHERE
              kc.status != "DELETED" AND
              NOT EXISTS (
                SELECT *
                FROM NODEPOOL np
                LEFT JOIN APP a ON np.id = a.nodepoolId
                WHERE
                  kc.id = np.clusterId AND np.isDefault = 0 AND
                  (
                    (a.status != "DELETED" AND a.status != "ERROR") OR
                    (a.status = "DELETED" AND a.destroyedDate > now() - INTERVAL 1 HOUR) OR
                    (a.status = "ERROR" AND a.dateAccessed > now() - INTERVAL 1 HOUR)
                  ));
```

Also want to do unit testing...